### PR TITLE
Enable kubeadm to configure auditing webhook

### PR DESCRIPTION
**Pull request by hh**
**What this PR does / why we need it**:

kubeadm should be able to configure an [audit webook backend](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/#webhook-backend).

**Which issue(s) this PR fixes**

kubernetes/kubeadm/issues/759

**Special notes for your reviewer**:

First k8s PR, guidance welcome.

**Release note**:

```release-note
Enable kubeadm configuration of Auditing Webhook Backend
```

### CODE REVIEW

File: cmd/kubeadm/app/cmd/init.go
Comment (chuckha) [inline cmd/kubeadm/app/cmd/init.go]: copy pasta error, should s/chuckha/hh

File: cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go
Comment (chuckha) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go]: This struct is starting to look like it wants to be two structs. One for the log backend configuration and one for the webhook configuration.

File: cmd/kubeadm/app/util/audit/utils.go
Comment (chuckha) [inline cmd/kubeadm/app/util/audit/utils.go]: I might have missed something but I think you can reuse the `writePolicyToDisk` function as the only thing I can see that has changed is the name of the variable `policyFile` variable?

File: cmd/kubeadm/app/util/audit/utils.go
Comment (chuckha) [inline cmd/kubeadm/app/util/audit/utils.go]: where is `name` defined?

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (chuckha) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: Watch out for this nil pointer dereference error -- the LogMaxAge has a use as a 0 value so it has to be a pointer (which would become nil) to indicate if no value was set to get a default value that is not 0.

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (chuckha) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: Please add to the test cases or create new test cases for the api server params

**Review by chuckha — CHANGES_REQUESTED** [review]
I need to find out if both backends can be enabled at the same time.

We should also think about splitting up the AuditPolicyConfiguration into two configurations. Might be out of scope for this ticket though.

File: cmd/kubeadm/app/util/audit/utils.go
Comment (hh) [reply cmd/kubeadm/app/util/audit/utils.go]: I need to refactor a bit, the Policy is of a type policy, I think an AuditWebhookConfig is of [type Config](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go#L53)

From https://kubernetes.io/docs/tasks/debug-application-cluster/audit/#use-logstash-to-collect-and-distribute-audit-events-from-webhook-backend:

```yaml
apiVersion: v1
kind: Config
clusters:
- cluster:
    server: http://<ip_of_logstash>:8888
  name: logstash
contexts:
- context:
    cluster: logstash
    user: ""
  name: default-context
current-context: default-context
preferences: {}
users: []
```

File: cmd/kubeadm/app/util/audit/utils.go
Comment (hh) [reply cmd/kubeadm/app/util/audit/utils.go]: Should have been webhookConfigFile

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (hh) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: Apiserver has a default built in value for --audit-webhook-initial-backoff, my intention is to omit the argument if not provided. I'll take a closer look at this today.

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (hh) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: Will do.

File: cmd/kubeadm/app/util/audit/utils.go
Comment (hh) [inline cmd/kubeadm/app/util/audit/utils.go]: Need to figure out how to register type v1.Config:

```error creating default audit webhook config file "/etc/kubernetes/audit/webhook.yaml" [failed to marshal audit webhook config to YAML: no kind is registered for the type v1.Config]```

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (hh) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: Fixes kubernetes/kubeadm/issues/765 !

**Comment by hh** [comment]
Must kubeadm + apiserver versions match?
If so, we should not need logic to decide which args are deprecated.

https://github.com/kubernetes/apiserver/commit/9169f6d3004339cca548211658891b7a8f347e3e#r28730299

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (hh) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: Needs to be moved into it's own PR

**Comment by rohfle** [comment]
I have signed the CLA as requested.

**Comment by timothysc** [comment]
This PR has way too many commits, please squash to 1 or 2 when you are ready for review. 

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews

**Comment by hh** [comment]
@chuckha do we need to do anything for our bazel job to start?

**Comment by chuckha** [comment]
/retest

they are probably stuck

**Comment by chuckha** [comment]
:/ @timothysc can you see why these are stuck? i can't retest

File: cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go
Comment (chuckha) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go]: Do either of these have default you want to set in v1alpha1/defaults.go?

**Review by chuckha — APPROVED** [review]
looks good

**Comment by chuckha** [comment]
To confirm my understanding: There is no default webhook.yaml file that we provide, probably because that's not something we can possibly provide a default for. Is that right?

**Comment by hh** [comment]
That's correct. It's format is the same as KUBECONFIG, giving an endpoint url to send the audit logs to.

File: cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go
Comment (hh) [reply cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go]: Neither of them need a default.

The webhook config file must be provided.

Initial backoff has a default of 10 set in [apiserver/.../webhook.go](https://github.com/ii/kubernetes/blob/a491ee569c0bf413d57be380535dedac9ade2724/staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go#L42)

```go
	// DefaultInitialBackoff is the default amount of time to wait before
	// retrying sending audit events through a webhook.
	DefaultInitialBackoff = 10 * time.Second
```

**Review by fabriziopandini — COMMENTED** [review]
@hh sorry for the delay in providing my feedback

Everything seems ok to me; I`m proposing only a small refactoring for keeping the init workflow clean and simple, but I let to @chuckha the final word on it,

File: cmd/kubeadm/app/cmd/init.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/init.go]: The amount of audit initialization logic embedded in the init workflow is starting to get relevant. What about moving it in auditutil.InitAuditing or something?
@chuckha WDYT?

Comment (chuckha) [reply cmd/kubeadm/app/cmd/init.go]: I agree, this section is getting fairly somewhat unruly, but I also don't want to get in @hh's way. I think being able to enable webhook auditing is a pretty good feature. If @hh is up for it, the refactor would be fantastic, but if not, I would be ok with a ticket with priority/important-soon to refactor this section since it is indeed getting a bit large for this init section. It might even be a good help-wanted ticket for the kubecon session this week (maybe?).

**Comment by hh** [comment]
Rebased to one commit.

**Comment by hh** [comment]
@chuckha @fabriziopandini @timothysc can I get a retest?

File: cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go]: Please add `,omitempty`, so the output of the struct serialization will be cleaner

**Review by fabriziopandini — CHANGES_REQUESTED** [review]
Please fix the build error (might be `hack/update-bazel.sh` is enough) and address latest comment; afterwards this is ready for merge IMO

**Comment by luxas** [comment]
@timothysc do you think we should try to squeeze this into v1.11 or wait?

**Comment by timothysc** [comment]
@luxas this is worth getting in imo.

**Comment by hh** [comment]
Can I get a sponsor for kubernetes membership? (to be able to /retest and be assigned PR reviews)

File: cmd/kubeadm/app/phases/controlplane/volumes.go
Comment (luxas) [inline cmd/kubeadm/app/phases/controlplane/volumes.go]: Isn't this already the case?

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (luxas) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: Why are these duplicated?

File: cmd/kubeadm/app/constants/constants.go
Comment (luxas) [inline cmd/kubeadm/app/constants/constants.go]: Do we actually _need_ this subdir or could it be statically mounted right into `/etc/kubernetes`?

**Review by luxas — COMMENTED** [review]
General comment: I don't see why you can't do this already today with extraargs/extravolumes? I'd **really** not want to clutter our API more than what we already have. That said, as a stop-gap I might be okay with merging this now, but I can't promise we're gonna keep it in forever as-is.

**Comment by luxas** [comment]
@hh you need a couple of more PRs submitted for that I think https://github.com/kubernetes/community/blob/master/community-membership.md

File: cmd/kubeadm/app/constants/constants.go
Comment (rohfle) [reply cmd/kubeadm/app/constants/constants.go]: We're reusing the existing audit policy subdirectory.
see cmd/kubeadm/app/constants/constants.go:380
```go
func GetStaticPodAuditWebhookConfigFile() string {
     return filepath.Join(KubernetesDir, **AuditPolicyDir**, AuditWebhookConfigFile)
}
```

**Comment by timothysc** [comment]
After looking through this some more, I don't see why we need this vs. documenting a path using arg overrides.  

/hold

**Comment by rohfle** [comment]
@timothysc - Forgive me if I am wrong - if you use extraargs, how do you get the audit webhook config into the container? At the moment, the audit config (for logging on disk) is set to be mounted as a volume into kube-apiserver by kubeadm. Thats why there is a separate non-extraargs way of doing things.

**Comment by rohfle** [comment]
I've been told that volume mounts are possible with the `APIServerExtraVolumes []HostPathMount` option in MasterConfiguration. This plus extra-args would probably allow you to achieve the same result as this PR. So this PR doesn't appear to be needed.

https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go#L94

**Comment by luxas** [comment]
> This plus extra-args would probably allow you to achieve the same result as this PR

Yes, exactly. I think documenting this is better.
With the coming `v1alpha2` API I'm ripping out everything that isn't critically needed, so even though if we'd merge this now there would be a fair chance we'd remove/revert that later.

@hh could you please document this functionality instead?

**Comment by luxas** [comment]
I'm sorry to close this, but documenting the usage of `.APIServerExtraArgs` and `.APIServerExtraVolumes` is the way we're gonna go for this feature, instead of baking it into our API. Thanks for the work on this PR though, sorry to not be able to merge it

**Pull request closed without merging**
